File: morkato/attack.py
from __future__ import annotations
from .utils import NoNullDict, extract_datetime_from_snowflake
from .flags import Flags
from typing_extensions import Self
from datetime import datetime
from typing import (
  TYPE_CHECKING,
  SupportsInt,
  Optional
)
if TYPE_CHECKING:
  from .types import Attack as AttackPayload
  from .state import MorkatoConnectionState
  from .guild import Guild
  from .art import Art
class AttackFlags(Flags):
  UNAVOIDABLE: int
  INDEFENSIBLE: int
  AREA: int
  NOT_COUNTER_ATTACKABLE: int
  COUNTER_ATTACKABLE: int
  DEFENSIVE: int
  def unavoidable(self) -> bool: ...
  def indefensible(self) -> bool: ...
  def area(self) -> bool: ...
  def not_counter_attackable(self) -> bool: ...
  def counter_attackable(self) -> bool: ...
  def defensive(self) -> bool: ...
class Attack:
  def __init__(self, state: MorkatoConnectionState, guild: Guild, art: Art, payload: AttackPayload) -> None:
    self.state = state
    self.http = state.http
    self.guild = guild
    self.art = art
    self.id = int(payload["id"])
    self.from_payload(payload)
  def from_payload(self, payload: AttackPayload) -> None:
    self.name = payload["name"]
    self.name_prefix_art = payload["name_prefix_art"]
    self.description = payload["description"]
    self.banner = payload["banner"]
    self.poison_turn = payload["poison_turn"]
    self.burn_turn = payload["burn_turn"]
    self.bleed_turn = payload["bleed_turn"]
    self.poison = payload["poison"]
    self.burn = payload["burn"]
    self.stun = payload["stun"]
    self.bleed = payload["bleed"]
    self.damage = payload["damage"]
    self.breath = payload["breath"]
    self.blood = payload["blood"]
    self.flags = AttackFlags(payload["flags"])
  @property
  def created_at(self) -> datetime:
    return extract_datetime_from_snowflake(self)
  async def update(
    self, *,
    name: Optional[str] = None,
    name_prefix_art: Optional[str] = None,
    description: Optional[str] = None,
    resume_description: Optional[str] = None,
    banner: Optional[str] = None,
    poison_turn: Optional[int] = None,
    burn_turn: Optional[int] = None,
    bleed_turn: Optional[int] = None,
    poison: Optional[int] = None,
    burn: Optional[int] = None,
    bleed: Optional[int] = None,
    stun: Optional[int] = None,
    damage: Optional[int] = None,
    breath: Optional[int] = None,
    blood: Optional[int] = None,
    flags: Optional[SupportsInt] = None
  ) -> Self:
    kwargs = NoNullDict(
      name=name,
      name_prefix_art=name_prefix_art,
      description=description,
      resume_description=resume_description,
      banner=banner,
      poison_turn = poison_turn,
      burn_turn = burn_turn,
      bleed_turn = bleed_turn,
      poison = poison,
      burn = burn,
      bleed = bleed,
      stun = stun,
      damage=damage,
      breath=breath,
      blood=blood,
      flags=flags
    )
    if kwargs:
      payload = await self.http.update_attack(self.guild.id, self.id, **kwargs)
      self.from_payload(payload)
    return self
  async def delete(self) -> Self:
    payload = await self.http.delete_attack(self.guild.id, self.id)
    self.from_payload(payload)
    self.art._del_attack(self)
    return self

File: api/src/main/kotlin/morkato/api/model/attack/Attack.kt
package morkato.api.model.attack

import morkato.api.infra.repository.AttackRepository
import morkato.api.model.guild.Guild

import org.jetbrains.exposed.sql.ResultRow
import java.math.BigDecimal

class Attack(
  val guild: Guild,
  val id: Long,
  val name: String,
  val artId: Long,
  val namePrefixArt: String?,
  val description: String?,
  val banner: String?,
  val poisonTurn: BigDecimal,
  val burnTurn: BigDecimal,
  val bleedTurn: BigDecimal,
  val poison: BigDecimal,
  val burn: BigDecimal,
  val bleed: BigDecimal,
  val stun: BigDecimal,
  val damage: BigDecimal,
  val breath: BigDecimal,
  val blood: BigDecimal,
  val flags: Int
) {
  public constructor(guild: Guild, row: ResultRow) : this(guild, AttackRepository.AttackPayload(row)) {}
  public  constructor(guild: Guild, payload: AttackRepository.AttackPayload) : this(
    guild,
    payload.id,
    payload.name,
    payload.artId,
    payload.namePrefixArt,
    payload.description,
    payload.banner,
    payload.poisonTurn,
    payload.burnTurn,
    payload.bleedTurn,
    payload.poison,
    payload.burn,
    payload.bleed,
    payload.stun,
    payload.damage,
    payload.breath,
    payload.blood,
    payload.flags
  );
  fun update(
    name: String? = null,
    namePrefixArt: String? = null,
    description: String? = null,
    banner: String? = null,
    poisonTurn: BigDecimal? = null,
    burnTurn: BigDecimal? = null,
    bleedTurn: BigDecimal? = null,
    poison: BigDecimal? = null,
    burn: BigDecimal? = null,
    bleed: BigDecimal? = null,
    stun: BigDecimal? = null,
    damage: BigDecimal? = null,
    breath: BigDecimal? = null,
    blood: BigDecimal? = null,
    flags: Int? = null
  ) : Attack {
    val payload = AttackRepository.AttackPayload(
      guildId = this.guild.id,
      id = this.id,
      name = name ?: this.name,
      artId = this.artId,
      namePrefixArt = namePrefixArt ?: this.namePrefixArt,
      description = description ?: this.description,
      banner = banner ?: this.banner,
      poisonTurn = poisonTurn ?: this.poisonTurn,
      burnTurn = burnTurn ?: this.burnTurn,
      bleedTurn = bleedTurn ?: this.bleedTurn,
      poison = poison ?: this.poison,
      burn = burn ?: this.burn,
      bleed = bleed ?: this.bleed,
      stun = stun ?: this.stun,
      damage = damage ?: this.damage,
      breath = breath ?: this.breath,
      blood = blood ?: this.blood,
      flags = flags ?: this.flags
    )
    AttackRepository.updateAttack(
      guildId = this.guild.id,
      id = this.id,
      name = name,
      namePrefixArt = namePrefixArt,
      description = description,
      banner = banner,
      poisonTurn = poisonTurn,
      burnTurn = burnTurn,
      bleedTurn = bleedTurn,
      poison = poison,
      burn = burn,
      bleed = bleed,
      stun = stun,
      damage = damage,
      breath = breath,
      blood = blood,
      flags = flags
    )
    return Attack(this.guild, payload)
  }
  fun delete() : Attack {
    AttackRepository.deleteAttack(this.guild.id, this.id)
    return this
  }
}
